File: frontend/app/src/main/java/com/capstone/whereigo/MainActivity.java
package com.capstone.whereigo;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.hardware.display.DisplayManager;
import android.opengl.GLES20;
import android.opengl.GLSurfaceView;
import android.os.Bundle;
import android.os.Handler;
import android.speech.SpeechRecognizer;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;
import android.content.Intent;

import androidx.appcompat.app.AlertDialog;
import androidx.core.view.WindowCompat;


import androidx.activity.OnBackPressedCallback;
import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.capstone.whereigo.databinding.ActivityMainBinding;
import com.google.android.material.search.SearchBar;
import com.google.android.material.search.SearchView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        EdgeToEdge.enable(this);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ImageButton backButton = findViewById(R.id.back_button);
        backButton.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, SearchActivity.class);
            startActivity(intent);
            finish();
        });

        ImageButton settingsButton = findViewById(R.id.settings_button);
        settingsButton.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, SettingActivity.class);
            startActivity(intent);
            finish();
        });

        if (savedInstanceState == null) {
            showFloorInputDialog();
        }
    }

    private void showFloorInputDialog() {
        View dialogView = getLayoutInflater().inflate(R.layout.dialog_floor_input, null);
        EditText editText = dialogView.findViewById(R.id.edit_floor_input);

        new AlertDialog.Builder(this)
                .setTitle("현재 층 입력")
                .setView(dialogView)
                .setCancelable(false)
                .setPositiveButton("확인", (dialog, which) -> {
                    String input = editText.getText().toString().trim();
                    if (!input.isEmpty()) {
                        Toast.makeText(this, input + "층 선택됨", Toast.LENGTH_SHORT).show();

                        // HelloArFragment 로드
                        getSupportFragmentManager().beginTransaction()
                                .setCustomAnimations(
                                        R.anim.enter_anim,       // 새 Fragment 등장
                                        R.anim.exit_anim,        // 현재 Fragment 퇴장
                                        R.anim.pop_enter_anim,   // 뒤로갈 때 새 Fragment 등장
                                        R.anim.pop_exit_anim     // 뒤로갈 때 현재 Fragment 퇴장
                                )
                                .replace(R.id.fragment_container, new HelloArFragment())
                                .commit();
                    } else {
                        Toast.makeText(this, "층을 입력해주세요", Toast.LENGTH_SHORT).show();
                        showFloorInputDialog(); // 재실행
                    }
                })
                .show();
    }
}

File: frontend/app/src/main/java/com/capstone/whereigo/MainActivity.kt
package com.capstone.whereigo

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.speech.SpeechRecognizer
import SearchResultAdapter
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.ImageButton
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.capstone.whereigo.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private val RECORD_AUDIO_REQUEST_CODE = 100
    private var backPressedTime: Long = 0
    private val backPressInterval: Long = 1000

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val searchBar = binding.searchBar
        val searchView = binding.searchView
        searchView.setupWithSearchBar(searchBar)

        val searchMenu = R.menu.search_menu
        searchBar.inflateMenu(searchMenu)

        // 설정 버튼 → SettingsFragment 전환
        findViewById<ImageButton>(R.id.settings_button).setOnClickListener {
            supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_setting, SettingsFragment())
                .addToBackStack(null)
                .commit()
        }

        // 음성 검색 클릭
        searchBar.menu.findItem(R.id.action_voice_search).setOnMenuItemClickListener {
            if (checkAudioPermission()) {
                if (SpeechRecognizer.isRecognitionAvailable(this)) {
                    val dialog = VoiceRecordDialog()
                    dialog.show(supportFragmentManager, "VoiceRecordDialog")
                } else {
                    Toast.makeText(this, "음성 인식을 사용할 수 없습니다.", Toast.LENGTH_SHORT).show()
                }
            } else {
                requestAudioPermission()
            }
            true
        }

        val recyclerView = searchView.findViewById<RecyclerView>(R.id.search_result)

        // SearchBar와 SearchView 연결
        searchView.setupWithSearchBar(searchBar)

        // 더미 데이터
        val allResults = listOf(
            "미래관 445호", "미래관 447호", "미래관 449호",
            "미래관 444호", "미래관 425호", "미래관 415호", "미래관 405호"
        )

        // 검색어 입력 후 엔터 시 필터링 & 표시
        searchView.editText.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                val query = searchView.text.toString()
                if (query.isNotBlank()) {
                    val filtered = allResults.filter { it.contains(query, ignoreCase = true) }
                    recyclerView.visibility = View.VISIBLE
                    recyclerView.layoutManager = LinearLayoutManager(this)
                    recyclerView.adapter = SearchResultAdapter(filtered)
                    searchView.hide()
                }
                true
            } else {
                false
            }
        }

        // 뒤로가기 처리
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (searchView.isShowing) {
                    searchView.hide()
                } else {
                    val currentTime = System.currentTimeMillis()
                    if (currentTime - backPressedTime < backPressInterval) {
                        isEnabled = false
                        onBackPressedDispatcher.onBackPressed()
                    } else {
                        backPressedTime = currentTime
                    }
                }
            }
        })

        // AR 프래그먼트 초기 진입
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, HelloArFragment())
                .commit()
        }
        // 카메라 Pose 표시 on/off
        searchView.addTransitionListener { _, _, newState ->
            HelloArFragment.setCameraPoseVisibility(
                newState != com.google.android.material.search.SearchView.TransitionState.SHOWN
            )
        }

        setupWakeWordListener()
    }
    // 오디오 권한 확인
    private fun checkAudioPermission(): Boolean {
        return ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED
    }

    // 오디오 권한 요청
    private fun requestAudioPermission() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(Manifest.permission.RECORD_AUDIO),
            RECORD_AUDIO_REQUEST_CODE
        )
    }

    // 권한 요청 결과 처리
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == RECORD_AUDIO_REQUEST_CODE) {
            if ((grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)) {
                Toast.makeText(this, "음성 권한이 허용되었습니다. 다시 시도해주세요.", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, "음성 권한이 거부되었습니다.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun setupWakeWordListener() {
        if (!checkAudioPermission()) {
            requestAudioPermission()
            return
        }
    }
}
